### ISSUE TITLE
create_log_entries_from_revisions crashes if a model uses on_delete=models.PROTECT

### ISSUE THREAD

**Issue by higs4281** [wagtail/wagtail#6368]
### Issue Summary

We upgraded to 2.10.1 and found that the log-entry script started crashing on a page type that has foreign-key relations to [three Django models](https://github.com/cfpb/consumerfinance.gov/blob/master/cfgov/jobmanager/models/pages.py#L82-L99), using `on_delete=models.PROTECT`.

We had tested the script in 2.10 with success.

We found that the script will run if we change the on_delete strategy to `SET_NULL`, which is not ideal. 

Wondering if this is related to [modelcluster issue 127](https://github.com/wagtail/django-modelcluster/issues/127)?

Stack trace:

```
Traceback (most recent call last):
  File "./cfgov/manage.py", line 11, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/higginsw/.pyenv/versions/c15gov/lib/python3.6/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/Users/higginsw/.pyenv/versions/c15gov/lib/python3.6/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/higginsw/.pyenv/versions/c15gov/lib/python3.6/site-packages/django/core/management/base.py", line 323, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/higginsw/.pyenv/versions/c15gov/lib/python3.6/site-packages/django/core/management/base.py", line 364, in execute
    output = self.handle(*args, **options)
  File "/Users/higginsw/.pyenv/versions/c15gov/lib/python3.6/site-packages/wagtail/core/management/commands/create_log_entries_from_revisions.py", line 34, in handle
    current_revision_as_page = revision.as_page_object()
  File "/Users/higginsw/.pyenv/versions/c15gov/lib/python3.6/site-packages/wagtail/core/models.py", line 1977, in as_page_object
    return self.page.specific.with_content_json(self.content_json)
  File "/Users/higginsw/.pyenv/versions/c15gov/lib/python3.6/site-packages/wagtail/core/models.py", line 1830, in with_content_json
    obj = self.specific_class.from_json(content_json)
  File "/Users/higginsw/.pyenv/versions/c15gov/lib/python3.6/site-packages/modelcluster/models.py", line 272, in from_json
    return cls.from_serializable_data(json.loads(json_data), check_fks=check_fks, strict_fks=strict_fks)
  File "/Users/higginsw/.pyenv/versions/c15gov/lib/python3.6/site-packages/modelcluster/models.py", line 239, in from_serializable_data
    obj = model_from_serializable_data(cls, data, check_fks=check_fks, strict_fks=strict_fks)
  File "/Users/higginsw/.pyenv/versions/c15gov/lib/python3.6/site-packages/modelcluster/models.py", line 100, in model_from_serializable_data
    raise Exception("can't currently handle on_delete types other than CASCADE, SET_NULL and DO_NOTHING")
Exception: can't currently handle on_delete types other than CASCADE, SET_NULL and DO_NOTHING
```

* I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: no

### Technical details

* Python version: 3.6.9
* Django version: 2.2.13
* Wagtail version: 2.10.1
* modelcluster version: 5.0.2

**Comment by gasman** [comment]
It looks like this can happen as a result of the following sequence of events:

* define a page model with a ForeignKey with on_delete=PROTECT
* save the page with the foreign key pointing to "item 1"
* resave the page with the foreign key pointing to "item 2"
* delete item 1

The failure occurs because it's not possible to instantiate the old revision as a valid page object once item 1 has been deleted - you'd see the same thing if you tried reverting to the old revision in the Wagtail admin's version management UI.

While this currently shows up as a "not implemented" exception in django-modelcluster, the "logically correct" version would be no better - it would just be an IntegrityError instead. I don't see any changes to this logic in the 2.10.1 version, so I can't see how this would have succeeded in 2.10 either. (I don't believe that https://github.com/wagtail/django-modelcluster/issues/127 is related - that one is specific to ParentalKey relations, where both sides of the relation are part of the JSON structure being unpacked.)

Since the only reason we're instantiating old revisions is to run a comparison against the previous revision and skip creating a log entry if they're identical, I think the best way around this is to sidestep the `as_page_object` calls entirely and just do the comparison at the JSON dict level. We'd need to consider the possibility of false positives where two JSON structures differ but don't have an actual content change, but the only case of that I can think of is differing tree `path` / `depth` values (and off the top of my head I'm not sure we even save those in revisions). As a bonus, it'll probably be a fair bit quicker to process. What do you think @zerolab?

**Comment by zerolab** [comment]
@gasman the management command does indeed compare the two revisions, using the baked in comparison methods. One thing that changes between revisions, and hence saved JSON, is the StreamField id even when there are no changes to it. If we go about comparing the JSON, we will end up with log entries for all revisions.

#6264 adds the first step in the path for avoiding new revisions being created without any content changes, but to replicate that we'd need to ignore StreamField ids and this then becomes a much bigger task than catching all exceptions 🤔

**Comment by higs4281** [comment]
> how this would have succeeded in 2.10

I think we fooled ourselves here. We tested, but I suspect it may have been deemed to be working when in fact it failed after running for some time.

**Comment by gasman** [comment]
> One thing that changes between revisions, and hence saved JSON, is the StreamField id even when there are no changes to it.

@zerolab Where have you seen that happening? Have just tested on bakerydemo, and IDs are being preserved as expected - I made two draft edits to a blog page, and StreamField block IDs were the same in both revisions, even when the content was changed. It would be concerning if that isn't the case, because tracking blocks across revisions is pretty much the whole purpose of those IDs...

**Comment by zerolab** [comment]
@gasman I went back to my notes and also retested. The StreamField IDs stay the same in the revision content. Sorry for the false alarm! `form.has_changes()` would trip up on StreamFields but that was fixed in #6264.

So the direct `content_json` comparison will work. 
The one thing that changes between two revisions is the value of `latest_revision_created_at`, that is outside of the path/depth (which we do save in content_json ).

**Comment by gasman** [comment]
Unfortunately it looks like there's one place in content_json that can introduce spurious differences and can't be trivially ignored: the ParentalKey on an inline child model is null up to and including the first published revision, and populated with the page ID in the following revision.

    # published revision
    "blog_person_relationship": [
      {
        "pk": 2,
        "sort_order": 0,
        "page": null,
        "people": 1
      }
    ],

    # subsequent saved draft with no other edits
    "blog_person_relationship": [
      {
        "pk": 2,
        "sort_order": 0,
        "page": 62,
        "people": 1
      }
    ],

Work in progress: https://github.com/gasman/wagtail/tree/fix/create-log-entries-json-comparison
